modules/core/src/main/java/org/apache/ignite/internal/processors/datastructures/GridCacheLockImpl2Unfair.java

```diff
@@ -0,0 +1,875 @@
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.lang.IgniteInClosure;
+import org.apache.ignite.plugin.extensions.communication.MessageReader;
+import org.apache.ignite.plugin.extensions.communication.MessageWriter;
+import org.jetbrains.annotations.Nullable;
+
+import static org.apache.ignite.internal.managers.communication.GridIoPolicy.P2P_POOL;
+
+/**
+ * Unfair implementation of reentrant lock based on a {@link IgniteInternalCache#invoke}.
+ */
+public final class GridCacheLockImpl2Unfair extends GridCacheLockEx2 {
+    /** Reentrant lock name. */
+    private final String name;
+
+    /** Cache context. */
+    private final GridCacheContext<GridCacheInternalKey, GridCacheLockState2Base<UUID>> ctx;
+
+    /** Internal synchronization object. */
+    private final LocalSync sync;
+
+    /** {@code ReleasedMessage} handler. */
+    private final IgniteInClosure<GridCacheIdMessage> releaser;
+
+    /**
+     * Constructor.
+     *
+     * @param name Reentrant lock name.
+     * @param key Reentrant lock key.
+     * @param lockView Reentrant lock projection.
+     */
+    GridCacheLockImpl2Unfair(String name, final GridCacheInternalKey key,
+        IgniteInternalCache<GridCacheInternalKey, GridCacheLockState2Base<UUID>> lockView) {
+        assert name != null;
+        assert key != null;
+        assert lockView != null;
+
+        this.name = name;
+        this.ctx = lockView.context();
+
+        final Latch listener = new Latch();
+
+        sync = new LocalSync(new GlobalSync(ctx.localNodeId(), key, lockView, listener, ctx,
+            ctx.logger(getClass())));
+
+        releaser = new IgniteInClosure<GridCacheIdMessage>() {
+            @Override public void apply(GridCacheIdMessage message) {
+                assert message instanceof ReleasedMessage;
+                assert key.name().equals(((ReleasedMessage)message).name);
+
+                listener.release();
+            }
+        };
+    }
+
+    /** {@inheritDoc} */
+    @Override void onNodeRemoved(UUID id) {
+        assert id != null;
+
+        sync.globalSync.remove(id);
+    }
+
+    /** {@inheritDoc} */
+    @Override public IgniteInClosure<GridCacheIdMessage> getReleaser() {
+        return releaser;
+    }
+
+    /** Message from previous node which release the lock. */
+    public static final class ReleasedMessage extends GridCacheIdMessage {
+        /** */
+        private static final long serialVersionUID = 181741851451L;
+
+        /** Message index. */
+        private static final int CACHE_MSG_IDX = nextIndexId();
+
+        /** Lock name. */
+        String name;
+
+        /**
+         * Empty constructor required for {@link Externalizable}.
+         */
+        public ReleasedMessage() {
+            // No-op.
+        }
+
+        /**
+         * Constructor.
+         *
+         * @param id Cache id.
+         * @param name Lock name.
+         */
+        ReleasedMessage(int id, String name) {
```

Comment:
public

Reply:
Why? I create ReleasedMessage only from this class. I have the public empty constructor for serialization.

Reply:
ok